skfda/preprocessing/dim_reduction/_pace.py

```diff
@@ -0,0 +1,1441 @@
+            None otherwise, and the search range if the bandwidth is a tuple,
+            or None otherwise. In the case that the bandwidth is None, the
+            function returns None, (0.1, 10.0) as the default search range.
+        """
+        if isinstance(bandwidth, float) and bandwidth <= 0:
+            error_msg = "Given bandwidth values must be positive."
+            raise ValueError(error_msg)
+
+        tuple_length = 2
+
+        if isinstance(bandwidth, Sequence) and (
+            len(bandwidth) != tuple_length or (
+                not all(isinstance(b, (float, int)) for b in bandwidth)
+            ) or bandwidth[0] <= 0 or bandwidth[1] <= bandwidth[0]
+        ):
+            error_msg = (
+                "Bandwidth search ranges must be a non-decreasing 2-sequence "
+                "of floats."
+            )
+            raise ValueError(error_msg)
+
+        if isinstance(bandwidth, float):
+            return bandwidth, None
+
+        return None, (bandwidth[0], bandwidth[1])
+
+    def __init__(  # noqa: PLR0913
+        self,
+        *,
+        n_components: float | None = None,
+        assume_noisy: bool = True,
+        kernel_mean: KernelFunction = gaussian_kernel,
+        bandwidth_mean: float | NDArrayFloat,
+        kernel_cov: KernelFunction = gaussian_kernel,
+        bandwidth_cov: float | NDArrayFloat,
+        bw_cov_n_grid_points: int = 30,
+        n_grid_points: int = 51,
+        boundary_effect_interval: Sequence[float] = (0.0, 1.0),
+        variance_error_interval: Sequence[float] = (0.25, 0.75),
+    ) -> None:
+        if n_components is None:
+            n_components = 1.0
+        else:
+            int_leq_0 = (isinstance(n_components, int) and n_components <= 0)
+            float_out_range = (
+                not isinstance(n_components, int) and (
+                    n_components <= 0.0 or n_components >= 1.0
+                )
+            )
+            if int_leq_0 or float_out_range:
+                error_msg = (
+                    "n_components must be an integer or a float in (0.0, 1.0)."
+                )
+                raise ValueError(error_msg)
+
+        bandwidth_mean_, bandwidth_mean_interval_ = self._check_bandwidth(
+            bandwidth_mean,
+        )
+
+        bandwidth_cov_, bandwidth_cov_interval_ = self._check_bandwidth(
+            bandwidth_cov,
+        )
+
+        if n_grid_points <= 0 or bw_cov_n_grid_points <= 0:
+            error_msg = "Grid points must be positive or None."
+            raise ValueError(error_msg)
+
+        tuple_length = 2
+        boundary_incorrect = (
+            len(boundary_effect_interval) != tuple_length or (
+                boundary_effect_interval[0] < 0
+            ) or (
+                boundary_effect_interval[1] > 1
+            ) or (
+                boundary_effect_interval[0] >= boundary_effect_interval[1]
+            )
+        )
```

Comment:
It is an easy to miss feature of Python, but comparison operators can be chained like in math notation:

```suggestion
        boundary_incorrect = (
            len(boundary_effect_interval) != tuple_length or not (
                0 < boundary_effect_interval[0] <= boundary_effect_interval[1] < 1
            ) 
        )
```